### PR TITLE
Fx: Integrate outbounds with peer lists

### PR DESCRIPTION
**Pull request by amckinney**
Now that we have a basic scaffold of the Fx packages, this introduces an initial integration between the outbound and peer list configurations.

At a high-level, we need to be able to construct a `yarpc.Chooser` that can be fed into each of the outbound constructors, e.g. `yarpchttp.Outbound`. The `yarpc.Chooser` interface is implemented by the peer list packages, such as `yarpcroundrobin`. And in order to construct a `yarpc.Chooser`, we need access to a `yarpc.Dialer`.

Given the above constraints, we decouple each of these components into separate `[Chooser|Dialer]Provider` abstractions so that we can have each of the transport modules producing `yarpc.Dialer`s and depending on `yarpc.Chooser`s.

The `yarpc.[Chooser|List]` implementations are driven by configuration, such that they are referenced under their respective transport, e.g. the `yarpc.[Dialer|Chooser]` is referenced under `yarpc.http.outbounds`, and they are defined under `yarpc.peers`, such as in `yarpc.peers.roundrobin`.

We need a way to map these definitions with their references, so the concept of a `Name` arises. The `yarpc.Dialer`s can be shared amongst all of the outbounds within a specific transport, so this proposes a simple `Name` method on the `yarpc.Dialer` interface that can be used to distinguish the `yarpc.Dialer`s from one another.

In the `yarpc.[Chooser|List]` case, there can be multiple peer lists for the same policy, e.g. `foo-round-robin` vs `bar-round-robin`. So in this case, we also define the `Name` method on the `yarpc.[Chooer|List]` types.

Below demonstrates an example configuration that would coordinate with the change introduced here:
```
yarpc:
  http:
    outbounds:
      foo:
        policy: foo-round-robin
      bar:
        policy: bar-round-robin

  peers:
    round-robin:
      foo-round-robin:
        capacity: 50
      bar-round-robin:
        capacity: 100      
```

I've attempted to make each commit reviewable on its own. Tests might not pass between commits: they are separated into logical chunks with an easier reviewing experience in mind.

### CODE REVIEW

File: v2/yarpchttpfx/yarpchttpfx.go [hunk 1]
```diff
@@ -113,7 +116,13 @@ type OutboundsConfig struct {
 
 // OutboundConfig is the configuration for constructing a specific outbound.
 type OutboundConfig struct {
+	// Specifies the address to use for this Outbound.
 	Address string `yaml:"address"`
+
+	// Specifies the peer list policy to use for this Outbound.
+	//
+	// If set, an address does not need to be configured.
+	Policy string `yaml:"policy"`
```
Comment (AlexAPB) [inline v2/yarpchttpfx/yarpchttpfx.go]: We spec-ed out `to` initially, but `policy` seems fine.

cc @kriskowal

**Comment by kriskowal** [comment]
I like @AlexAPB’s suggestion to strike public Register methods from the record.

File: v2/yarpchttpfx/yarpchttpfx.go (hunk 1 above)
Comment (AlexAPB) [reply v2/yarpchttpfx/yarpchttpfx.go]: this is fine as is, just wanted to call it out

**Comment by amckinney** [comment]
Discussed offline - 
Seeing as this is a finicky test with the thriftrw plugin, merging this as is.

**Comment by amckinney** [comment]
Thanks @kriskowal I'll address the outstanding concerns in a follow-up!

**Pull request merged**